wremnants/muon_calibration.py

```diff
@@ -68,5 +89,63 @@ def define_corrected_muons(df, helper):
     df = df.Define("Muon_correctedEta", "ROOT::VecOps::RVec<float> res(Muon_correctedMom4Charge.size()); std::transform(Muon_correctedMom4Charge.begin(), Muon_correctedMom4Charge.end(), res.begin(), [](const auto &x) { return x.first.Eta(); } ); return res;")
     df = df.Define("Muon_correctedPhi", "ROOT::VecOps::RVec<float> res(Muon_correctedMom4Charge.size()); std::transform(Muon_correctedMom4Charge.begin(), Muon_correctedMom4Charge.end(), res.begin(), [](const auto &x) { return x.first.Phi(); } ); return res;")
     df = df.Define("Muon_correctedCharge", "ROOT::VecOps::RVec<int> res(Muon_correctedMom4Charge.size()); std::transform(Muon_correctedMom4Charge.begin(), Muon_correctedMom4Charge.end(), res.begin(), [](const auto &x) { return x.second; }); return res;")
+    return df
 
+def get_good_gen_muons_idx_in_GenPart(df, reco_subset = "goodMuons"):
+    df = df.Define("goodMuons_idx", f"Muon_genPartIdx[{reco_subset}]")
+    df = df.Define("goodMuonsByGenTruth",
+        (
+        "ROOT::VecOps::RVec<bool> res(goodMuons_idx.size());"
+        "for (int i = 0; i < goodMuons_idx.size(); i++) {"
+        "    res[i] = ("
+        "        (GenPart_statusFlags[goodMuons_idx[i]] & 0x01) &&"
+        "        (abs(GenPart_pdgId[goodMuons_idx[i]]) == 13) &&"
+        "        (GenPart_status[goodMuons_idx[i]] == 1)"
+        "    );"
+        "}"
+        "return res;"
+        )
+    )
+    df = df.Define("goodGenMuons_idx","goodMuons_idx[goodMuonsByGenTruth]")
+    df = df.Filter("goodGenMuons_idx.size() > 0")
     return df
+
+def define_good_gen_muons_kinematics(df, kinematic_vars = ["pt", "eta", "phi", "charge"]):
+    for var in kinematic_vars:
+        if var == "charge": var = "pdgId"
+        col_name = f"GenMuon_{var}"
+        selection = f"ROOT::VecOps::Take(GenPart_{var}, goodGenMuons_idx)"
+        df = df.Define(col_name, selection)
+        if var == "pdgId":
+            df = df.Define("GenMuon_charge", 
+                (
+                "ROOT::VecOps::RVec<int> res(GenMuon_pdgId.size());"
+                "for (int i = 0; i < GenMuon_pdgId.size(); i++) {"
+                "res[i] = GenMuon_pdgId[i] > 0 ? -1 : 1;}"
+                "return res;"
+                )
+            )
+    return df
+
+def transport_smearing_weights_to_reco(
+    resultdict, procs = ['WplusmunuPostVFP', 'WminusmunuPostVFP', 'ZmumuPostVFP'],
+    proj_axes = ['pt', 'eta']
+):
+    for proc in procs:
+        proc_hists = resultdict[proc]['output']
+        if any(x in proc for x in ['plus', 'Plus']): charge = 1
+        elif any(x in proc for x in ['plus', 'Plus']): charge = -1
```

Comment:
Is this meant to be minus? You could also just check "plus" in proc.lower()

Reply:
oops, thanks for catching that, fixed